### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by SzajekSzymon**
## Description
Relates to #51 .

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -262,30 +262,127 @@ Parser consists of the following phases:
 ## Used technologies
 
 ### Extensions / Plugins
+main plugins:
+
+* React - https://reactjs.org/
```
Comment (tomaszmichalak) [inline README.md:267]: maybe simply:

```
* [React](https://reactjs.org/)
* [Redux](https://redux.js.org/)
* [Webext](https://github.com/tshaddix/webext-redux)
* [styled-components](https://styled-components.com/)
* [vis-network](https://visjs.github.io/vis-network/docs/network/)
* [vis-timeline](https://visjs.github.io/vis-timeline/docs/timeline/)
* [unique-selector](https://github.com/ericclemmons/unique-selector)
* [renderjson](https://github.com/caldwell/renderjson)
* [react-fontawesome](https://github.com/FortAwesome/react-fontawesome)
* [Jest](https://jestjs.io/docs/en/cli.html)
* [Enzyme](https://enzymejs.github.io/enzyme/)
* [semver](https://github.com/npm/node-semver#readme)
* [Babel](https://babeljs.io/)
* [Eslint](https://eslint.org/)
* [Webpack](https://webpack.js.org/)
```

Can we divide them in categories:
- application / components
- testing / quality / code conventions
- delivery / deployment

File: README.md [hunk 2]
```diff
@@ -262,30 +262,127 @@ Parser consists of the following phases:
...
+* react-fontawesome - https://github.com/FortAwesome/react-fontawesome
+* Jest - https://jestjs.io/docs/en/cli.html
+* Enzyme - https://enzymejs.github.io/enzyme/
+* semver - https://github.com/npm/node-semver#readme
+* Babel - https://babeljs.io/
+* Eslint - https://eslint.org/
+* Webpack - https://webpack.js.org/
 
 ### CI
+We use CI to provide a highest quality of our code.
```
Comment (tomaszmichalak) [inline README.md:284]: Maybe something like:

> The GitHub repository is integrated with [Azure Pipelines](https://dev.azure.com/knotx/Knotx/_build?definitionId=16&_a=summary) (CI) to validate both new PRs and the master branch. Check the `azure-pipelines.yml` file for configuration details.  So we  check:
> - code conventions with Eslint
> - code logic with unit tests using Jest
> - test coverage level with preconfigured thresholds (see `jest.config.js` for more details).

File: README.md [hunk 3]
```diff
@@ -262,30 +262,127 @@ Parser consists of the following phases:
...
+
+CI checks:
+* Tests passing
+* App build without fails
+* Coverage level
+
+
+## Testing
+Currently we test our app only through the unit tests. To testing we use two main technologies: Jest + Enzyme.
```
Comment (tomaszmichalak) [inline README.md:293]: We believe that unit tests remain the best documentation. All React components, processing logic (helpers) and actions (such as a button click) are validated with unit tests. We use `Jest` and `Enzyme` frameworks to validate both components (React) with combination with mocked storage (Redux). 

All JS files (components & helpers) have their own tests that are placed next to the tested sources. We follow the convention:
`*.mock.js` - it is configuration containing mocks for our tests
`*.spec.jsx` - it contains unit tests

Additionally, we placed tests coverage verification in our CI. We use the `jest-coverage` tool for that. We decided to keep the coverage level at truly high levels (80 - 100%). It should enable future refactoring and code changes.

When tests are executed, then we generate the report (`test-report.xml`) file in the `build/test` folder. Moreover, there is the `coverage` directory that contains the `intex.html` file with unit tests coverage report.

File: README.md [hunk 4]
```diff
@@ -262,30 +262,127 @@ Parser consists of the following phases:
...
+└── exampleHelper.spec.js
+```
+
+We use jest-coverage tool. After run tests in the root directory jest create a build directory. In this folder you can find an index.html file with coverage report. We try to stay on 80 - 100% coverage level.
+
+```
+build/test/coverage/index.html
+```
+
+### How to test?
```
Comment (tomaszmichalak) [inline README.md:310]:
```suggestion
### How to run tests?
```

File: README.md [hunk 5]
```diff
@@ -262,30 +262,127 @@ Parser consists of the following phases:
...
+1. run command to fire all tests: `yarn run test`
+2. run command to fire specific test: `yarn run test [path_to_test]`
```
Comment (tomaszmichalak) [inline README.md:312]:
```suggestion
2. run command to fire the specific test: `yarn run test [path_to_test]`
```

File: README.md [hunk 6]
```diff
@@ -262,30 +262,127 @@ Parser consists of the following phases:
...
 
 ## Implementation details
 
-### Parsing HTML
+### Data flow
+Extension get information about knot.x from html markup. Knot.x in ebug mode provide a tag with all necessary data. In the next step our content script get this data and change structure to comfortable for us. Content script pass this data to background script which save them in redux.
```
Comment (tomaszmichalak) [inline README.md:317]: > An extension's architecture includes such components as:
> - `background script` that contains listeners for browser events and communication with `Redux`
> - `content script` that contains JavaScript that executes in the context of a page that has been loaded in the browser.
>
> Depending on the format (JSON or HTML markup), fragments' debug data is processed by the content script first and then stored in `Redux` with the background script (which basically wraps Redux store). Content script transformations are described [here](https://github.com/Knotx/knotx-fragments-chrome-extension#debug-fragment-data).

File: README.md [hunk 7]
```diff
@@ -262,30 +262,127 @@ Parser consists of the following phases:
...
+
+```
+KNOT.x -> HTML MARKUP -> CONTENT SCRIPT -> BACKGROUND SCRIPT -> REDUX -> COMPONENTS
+                               ^
+                               |
+                         Parsing helper
+```
 
 ### Components
+The components structure in the main concept look like this:
```
Comment (tomaszmichalak) [inline README.md:327]: @Voycawojka do we have some tools to render this structure automatically?

Comment (tomaszmichalak) [reply README.md:327]: Check this one please: https://github.com/doczjs/docz

File: README.md [hunk 8]
```diff
@@ -262,30 +262,127 @@ Parser consists of the following phases:
...
+    └── MainPanel
+        └── Graph
+            ├──  Timeline
+            ├──  Legend
+            │    └── LegendSection
+            └──  NodeInfo
+```
+
+#### Graph && timelines
+To create a data visualization we use vis.js library. Currently we have 3 charts in our application.
```
Comment (tomaszmichalak) [inline README.md:347]:
```
We use `vis.js` library to visualise [fragments](https://github.com/Knotx/knotx-fragments/tree/master/api#fragment)' processing details. See the following components:
- Timeline showing the processing time of  all fragments (`SidePanel `: `FragmentGantt` component)
- Chart presenting the logic of processing a particular fragment (`MainPanel `: `Graph` component)
- Timeline showing the processing times of all steps performed while processing a specific fragment (`MainPanel `: `Timeline` component)
```
- Please rename one component:  `FragmentGannt` -> `FragmentGantt`.

File: README.md [hunk 9]
```diff
@@ -262,30 +262,126 @@ Parser consists of the following phases:
...
+
+some global styling and styling for renderjson markupwe store in
+ ```
+/src/js/styling/globalStyle.js
+  ```
 
-#### Timelines
 
 ### Data storage
+All data about fragments and pages are store in redux. We store a separate data set and an app state for all active pages to allow using a few dev tools console for a few Knot.x pages. Every pageData-set has own ID (chrome tab id). The structure of redux looks like this:
```
Comment (tomaszmichalak) [inline README.md:370]:
```
We use Redux as storage. It keeps details about:
parsed list of fragments
application state such as details which panel was expanded/hidden etc.
Once loaded page data is stored in a map where:
 a key is a Chrome tab identifier 
a value contains fragments, page data and application state per tab.
Such storage solution makes it easy to analyse many pages at the same time, switching between them, and running many Chrome Dev Tools Console instances.

The example below presents how data is stored in Redux:
```

File: README.md [hunk 10]
```diff
@@ -262,30 +262,126 @@ Parser consists of the following phases:
...
+    ├── 78: // tab id
+    │   ├── fragments: [] // list of fragments
+    │   ├── url: "https://example.com // page url
+    │   ├── sidebarExpanded: true // side panel expanded switch
+    │   └── renderedGraph: null // id of the currently selected fragment
+    └── 110:
+        └── ...
+```
+
+The pageData item is created on page load and destroy on page close. We create an item for all active pages. If the page doesn't use Knot.x , fragments property is empty.
```
Comment (tomaszmichalak) [inline README.md:385]: The pageData entry is created on page load and destroyed when we close the tab. If the page does not contain Knot.x fragments, `fragments` property is empty.

File: README.md [hunk 11]
```diff
@@ -262,30 +262,126 @@ Parser consists of the following phases:
...
+#### Graph && timelines
+To create a data visualization we use vis.js library. Currently we have 3 charts in our application.
+
+* Timeline for fragments on sidePanel (FragmentGannt component)
+* Main graph for the fragment (Graph component)
+* Performance timeline for the fragment (Timeline component)
 
-#### Graph
+### Styling
+We don't use any grid system to make our app beautiful. Everything is flex. To show and hide elements we try to use a react state, without saving this information in the redux store.
```
Comment (tomaszmichalak) [inline README.md:353]: It is a bit not consistent: we write that we try to use a react state, in storage section we describe that we have details about expanded panels.

Comment (SzajekSzymon) [reply README.md]: I added info about sidePanel exception. This is only component witch has an expanded info stored in redux.

**Pull request merged**